flink-tests/src/test/java/org/apache/flink/test/streaming/api/datastream/WatermarkITCase.java

```diff
@@ -0,0 +1,1064 @@
+            Map<Integer, Watermark> emitWatermarksInOperator2,
+            WatermarkHandlingResult operator3HandlingResult)
+            throws ReflectiveOperationException {
+        ExecutionEnvironmentImpl env =
+                (ExecutionEnvironmentImpl) ExecutionEnvironment.getInstance();
+        env.getConfiguration().set(PipelineOptions.OPERATOR_CHAINING, false);
+        ProcessConfigurableAndNonKeyedPartitionStream<Long> source =
+                env.fromSource(
+                                DataStreamV2SourceUtils.fromData(
+                                        LongStream.range(0, NUMBER_KEYS)
+                                                .boxed()
+                                                .collect(Collectors.toList())),
+                                "Operator1")
+                        .withParallelism(DEFAULT_PARALLELISM);
+
+        source.shuffle()
+                .process(
+                        new Operator2ProcessFunction(
+                                watermarkDeclaration, emitWatermarksInOperator2))
+                .withName("Operator2")
+                .withParallelism(DEFAULT_PARALLELISM)
+                .shuffle()
+                .process(new Operator3ProcessFunction(operator3HandlingResult))
+                .withName("Operator3")
+                .withParallelism(DEFAULT_PARALLELISM)
+                .shuffle()
+                .process(new Operator4ProcessFunction())
+                .withName("Operator4")
+                .withParallelism(DEFAULT_PARALLELISM);
+        return env.getStreamGraph();
+    }
+
+    public StreamGraph getStreamGraphForTestSource(
+            WatermarkDeclaration watermarkDeclaration,
+            Map<Integer, Watermark> emitWatermarksInSourceOperator1)
+            throws ReflectiveOperationException {
+        ExecutionEnvironmentImpl env =
+                (ExecutionEnvironmentImpl) ExecutionEnvironment.getInstance();
+        env.getConfiguration().set(PipelineOptions.OPERATOR_CHAINING, false);
+
+        ProcessConfigurableAndNonKeyedPartitionStream<Long> source =
+                env.fromSource(
+                                DataStreamV2SourceUtils.wrapSource(
+                                        new Operator1Source(
+                                                0,
+                                                NUMBER_KEYS - 1,
+                                                watermarkDeclaration,
+                                                emitWatermarksInSourceOperator1)),
+                                "Operator1")
+                        .withParallelism(DEFAULT_PARALLELISM);
+
+        source.shuffle()
+                .process(new Operator2ProcessFunction(null, null))
+                .withName("Operator2")
+                .withParallelism(DEFAULT_PARALLELISM);
+        return env.getStreamGraph();
+    }
+
+    public StreamGraph getStreamGraphForAlignedWatermark(List<Boolean> emitWatermarkValues)
```

Comment:
`emitWatermarkValues` is too difficult to understand. It's better to rename and replace it with map directly.